src/app/clusters/actions-server/actions-server.h

```diff
@@ -0,0 +1,108 @@
+namespace Actions {
+
+class Delegate;
+
+class Instance : public AttributeAccessInterface, public CommandHandlerInterface
+{
+public:
+    // Register for the Actions cluster on all endpoints.
+    Instance() :
+        AttributeAccessInterface(Optional<EndpointId>::Missing(), Actions::Id),
+        CommandHandlerInterface(Optional<EndpointId>::Missing(), Actions::Id)
+    {}
+
+    static Instance * GetInstance();
+    void SetDefaultDelegate(Delegate * aDelegate);
+
+private:
+    Delegate * mDelegate;
+    static Instance instance;
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+
+    CHIP_ERROR ReadActionListAttribute(const AttributeValueEncoder::ListEncodeHelper & encoder);
+    CHIP_ERROR ReadEndpointListAttribute(const AttributeValueEncoder::ListEncodeHelper & encoder);
+    CHIP_ERROR FindActionIdInActionList(uint16_t actionId);
+
+    // CommandHandlerInterface
+    template <typename RequestT, typename FuncT>
+    void HandleCommand(HandlerContext & handlerContext, FuncT func);
+
+    void InvokeCommand(HandlerContext & handlerContext) override;
+
+    void HandleInstantAction(HandlerContext & ctx, const Commands::InstantAction::DecodableType & commandData);
+    void HandleInstantActionWithTransition(HandlerContext & ctx,
+                                           const Commands::InstantActionWithTransition::DecodableType & commandData);
+    void HandleStartAction(HandlerContext & ctx, const Commands::StartAction::DecodableType & commandData);
+    void HandleStartActionWithDuration(HandlerContext & ctx, const Commands::StartActionWithDuration::DecodableType & commandData);
+    void HandleStopAction(HandlerContext & ctx, const Commands::StopAction::DecodableType & commandData);
+    void HandlePauseAction(HandlerContext & ctx, const Commands::PauseAction::DecodableType & commandData);
+    void HandlePauseActionWithDuration(HandlerContext & ctx, const Commands::PauseActionWithDuration::DecodableType & commandData);
+    void HandleResumeAction(HandlerContext & ctx, const Commands::ResumeAction::DecodableType & commandData);
+    void HandleEnableAction(HandlerContext & ctx, const Commands::EnableAction::DecodableType & commandData);
+    void HandleEnableActionWithDuration(HandlerContext & ctx,
+                                        const Commands::EnableActionWithDuration::DecodableType & commandData);
+    void HandleDisableAction(HandlerContext & ctx, const Commands::DisableAction::DecodableType & commandData);
+    void HandleDisableActionWithDuration(HandlerContext & ctx,
+                                         const Commands::DisableActionWithDuration::DecodableType & commandData);
+};
+
+class Delegate
+{
+public:
+    virtual ~Delegate() = default;
+
+    virtual CHIP_ERROR ReadActionAtIndex(uint16_t index, Structs::ActionStruct::Type & action);
+    virtual CHIP_ERROR ReadEndpointListAtIndex(uint16_t index, Structs::EndpointListStruct::Type & epList);
```

Comment:
Again, this is making lifetime assumptions.

Reply:
Added documentation.